runtime/compiler/env/PersistentAllocator.cpp

```diff
@@ -194,25 +254,49 @@ PersistentAllocator::freeBlock(Block * block)
    }
 
 void
-PersistentAllocator::deallocate(void * mem, size_t) throw()
+PersistentAllocator::freeBlock(Block * block)
    {
-   if (::memoryAllocMonitor)
-      ::memoryAllocMonitor->enter();
-
-
-   Block * block = static_cast<Block *>(mem) - 1;
+   TR_ASSERT(block->_size > 0, "Block size is non-positive");
 
-   // adjust the used persistent memory here and not in freePersistentmemory(block, size)
+   // Adjust the used persistent memory here and not in freePersistentmemory(block, size)
    // because that call is also used to free memory that wasn't actually committed
-   TR::AllocatedMemoryMeter::update_freed(block->_size, persistentAlloc);
+   if (TR::AllocatedMemoryMeter::_enabled & persistentAlloc)
+      {
+      j9thread_monitor_enter(_smallBlockListsMonitor);
+      TR::AllocatedMemoryMeter::update_freed(block->_size, persistentAlloc);
+      j9thread_monitor_exit(_smallBlockListsMonitor);
+      }
+  
+   // If this is a small block, add it to the appropriate fixed-size-block
+   // chain. Otherwise add it to the variable-size-block chain which is in
+   // ascending size order.
+   //
+   size_t const index = freeBlocksIndex(block->_size);
+   if (index > 0)
+      {
+      j9thread_monitor_enter(_smallBlockListsMonitor);
+      freeFixedSizeBlock(block);
+      j9thread_monitor_exit(_smallBlockListsMonitor);
+      }
+   else
+      {
+      if (::memoryAllocMonitor)
+         ::memoryAllocMonitor->enter();
+      freeVariableSizeBlock(block);
+      if (::memoryAllocMonitor)
+         ::memoryAllocMonitor->exit();
+      }
+   }
 
+void
+PersistentAllocator::deallocate(void * mem, size_t) throw()
+   {
+   Block * block = static_cast<Block *>(mem) - 1;
+   TR_ASSERT_FATAL(block->_next == NULL, "Freeing a block that is in use. block=%p next=%p", block, block->_next);
```

Comment:
Shouldn't the message say something like "Freeing a block that is in the free list"?

Reply:
Right. The message was meant to alert for double free which we see from time to time